### ISSUE TITLE
RavenDB timeout query can hang endpoint and can hang the shutdown process

### ISSUE THREAD

**Issue by SzymonPobiega** [Particular/NServiceBus.RavenDB#140]
Task force: @mauroservienti @DavidBoike @colin-higgins 

### Symptoms

Endpoint is unable to complete the shutdown procedure due to timeout poller being stuck in `GetNextChunk` query.

### Description

The RavenDB implementation of `GetNextChunk` uses the stream API which can block forever if inserts outpace the reads. In our scenario we are sending out messages (doing I/O operations) inside the loop so if the queue is slower than raven inserts, the query will never return.

Here's a snippet that shows the problem
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Raven.Client.Document;

namespace RavenNeverendingStream
{
    class Program
    {
        static void Main(string[] args)
        {
            var store = new DocumentStore()
            {
                Url = "http://localhost:8083"
            }.Initialize();

            var producer = Task.Run(() =>
            {
                var i = 0;
                while (true)
                {
                    using (var session = store.OpenSession("Test1"))
                    {
                        session.Store(new Timeout {Number = i});
                        session.SaveChanges();
                        i++;
                    }
                    Thread.Sleep(150);
                }
            });

            var consumer = Task.Run(() =>
            {
                while (true)
                {
                    try
                    {
                        Console.WriteLine("Starting new query.");
                        using (var session = store.OpenSession("Test1"))
                        {
                            Console.WriteLine("Opening stream.");
                            var query = session.Query<Timeout>("SomeIndex");
                            var stream = session.Advanced.Stream(query);
                            Console.WriteLine("Enumerating the stream.");
                            int j = 0;
                            while (stream.MoveNext())
                            {
                                j++;
                                //Console.WriteLine(stream.Current.Document.Id);
                                session.Delete(stream.Current.Document.Id);
                                Thread.Sleep(200);
                            }
                            Console.WriteLine(j);
                        }
                    }
                    catch (Exception ex)
                    {
                        Console.WriteLine(ex);
                    }
                    
                }
            });
            Console.ReadLine();
        }
    }

    public class Timeout
    {
        public string Id { get; set; }
        public int Number { get; set; }
    }
}
```

### Who's affected

RavenDB persistence users who have endpoints with high timeout throughput.

### Suggested resolution

Bound the stream query to not buffer more than a certain number of timeouts

**Comment by mauroservienti** [comment]
@SzymonPobiega something like `.Take( timeoutsCountToFetch )` when setting up the query?

**Comment by SzymonPobiega** [comment]
Yeah, exactly.

**Comment by mauroservienti** [comment]
@colin-higgins @DavidBoike let's discuss this in today sync. We should backport as well.

**Comment by andreasohlund** [comment]
Since we also have reports of the stream api getting stuck during normal operations as well I'd say we should prioritize and patch this one?

**Comment by andreasohlund** [comment]
We've decided to patch this one since it can stop firing timeouts if the new timeouts are added constantly to the storage.

@DavidBoike added this to the `3.1.1` milestone just to flag it as "being patched". If your fix is done before this one we'll just move this to a `3.1.2`

**Comment by andreasohlund** [comment]
@SzymonPobiega @mauroservienti @colin-higgins @DavidBoike did you discuss this?

We need to get this one going? (any takers)

**Comment by SzymonPobiega** [comment]
@andreasohlund I don't have capacity right now. Enrolled for the outbox performance test thing.

**Comment by ramonsmits** [comment]
So, here I also see that the stream API is used. Looks very similar to Particular/ServiceControl#650 Why isn't DeleteByIndex being used?

**Comment by mauroservienti** [comment]
So, the actual `GetChunkQuery` looks like:

```
session.Query<TimeoutData, TimeoutsIndex>()
    .OrderBy(t => t.Time)
    .Where( t =>
    {
        t.OwningTimeoutManager == string.Empty 
        || t.OwningTimeoutManager == EndpointName
    } )
    .Where(t => t.Time > startSlice)
    .Select(t => new
    {
        t.Id,
        t.Time
     } );
```

we then stream the above query stopping if the Document `Time` is later than `UctNow`, why not simply setting that condition in query itself?

```
session.Query<TimeoutData, TimeoutsIndex>()
    .OrderBy(t => t.Time)
    .Where( t =>
    {
        t.OwningTimeoutManager == string.Empty 
        || t.OwningTimeoutManager == EndpointName
    } )
    .Where(t => t.Time > startSlice && t.Time < DateTime.UtcNow )
    .Select(t => new
    {
        t.Id,
        t.Time
     } );
```

**Comment by mauroservienti** [comment]
I got it, the idea when using the streaming API is to avoid the pagination issue

**Comment by andreasohlund** [comment]
>I got it, the idea when using the streaming API is to avoid the pagination issue

Yes


Since the number of documents found can be large I'd say a cap on number is better since a cap on time might still endup with to many documents to dispatch?

**Comment by mauroservienti** [comment]
Well, the RavenDB client query is capped by default at 128 results, given that we can retrieve query statistics we have all the required information to decide if we need to immediately run another query or not.

I'll take a stab at this, assigning to myself.

**Comment by andreasohlund** [comment]
>Well, the RavenDB client query is capped by default at 128 results,

Hmm, are you saying that the stream will only return 128? ( @SzymonPobiega 's test didn't display that behavior)

**Comment by mauroservienti** [comment]
no, a standard query, a stream will return all the results, but it hangs on shutdown. So an option is to not use a stream that may not be really required here.

**Comment by mauroservienti** [comment]
Another option could be a complete redesign allowing extensions such as this to be notified of the endpoint shut down

**Comment by andreasohlund** [comment]
>Another option could be a complete redesign allowing extensions such as this to be notified of the endpoint shut down

You're after a `Feature.Stop()` method right?

@Particular/nservicebus-maintainers I think we should consider this for v6

@mauroservienti in the meantime you could use FeatureStartupTask.Stop as a stop gap?

**Comment by mauroservienti** [comment]
@andreasohlund 

> @mauroservienti in the meantime you could use FeatureStartupTask.Stop as a stop gap?

are you basically saying:

* create a `FeatureStartupTask` that has a reference to the `IQueryTimeouts`
* in the `Stop` method inject the stop request into the `IQueryTimeouts` implementation
* if the abort request is injected exit the streaming loop if it is running

Given that the `QueryTimeouts` that we have is registered as singleton it should work as expected. I'll set it up and see what happens.

**Comment by andreasohlund** [comment]
Yea. Isn't the polling already a startup task btw?

**Comment by mauroservienti** [comment]
I'll have a look

**Comment by mauroservienti** [comment]
changed the title and removed the milestone to identify that this is related to vNext

@Particular/platform-dev-squad this is missing a proper task force

**Comment by mauroservienti** [comment]
We had a call this morning regarding [the IQueryTimeouts cancellation token](https://github.com/Particular/NServiceBus/pull/3497) and we decided to drop it in favor of https://github.com/Particular/NServiceBus/issues/3532.

So the goal here is to do the minimum amount of required work to prevent the issue to affect customers:

* Use `FeatureStartupTask` to plugin shutdown logic
   * We have no evidence that users need to replace the `IQueryTimeouts` implementation so we don't really care, right now, if shutdown logic is bound to our internal implementation
* Find a way to prevent the streaming API unbounded usage when querying for timeouts

**Comment by adamralph** [comment]
*(Following https://github.com/Particular/Collaboration/blob/master/Staffing-Task-Forces.md.)*

Wwhat knowledge and skills are needed on the task force?

/cc @Particular/platform-dev-squad @Particular/persistence-maintainers

**Comment by mauroservienti** [comment]
updated with the proper task force

**Comment by mauroservienti** [comment]
One of the issues that we have is that there is a customer report stating that this can hang even if there are no timeouts at all, the possible reason is that the streaming API is hanging since the `TimeoutManager` is checking for timeouts every x seconds hitting the streaming API.
One option could be to entirely remove the dependency on the streaming API and issue a query capped at 1024 results.
We are using the [last document in the chunk](https://github.com/Particular/NServiceBus.RavenDB/blob/develop/src/NServiceBus.RavenDB/Timeouts/QueryTimeouts.cs#L62) to determine when to run the next query, that means that I see no way we can skip any timeout.

So my idea is to change the `GetNextChunk` implementation and make it behave as the `GetCleanupChunk` using a standard query with `.Take(1024)`.

@colin-higgins @DavidBoike comments?

**Comment by colin-higgins** [comment]
Yeah that sounds good to me, it will be a lot more simple to reason about as well.

**Comment by mauroservienti** [comment]
Can the [LongRunning](https://github.com/Particular/NServiceBus/blob/master/src/NServiceBus.Core/Timeout/Hosting/Windows/TimeoutPersisterReceiver.cs#L57) setting be an issue? We have a Desk case reporting that the endpoint hangs even if there are no timeouts at all.

**Comment by colin-higgins** [comment]
I converted raven to use the paging queries as discussed, which will prevent the theoretical scenario of timeouts being inserted faster than they can be pulled. 
I looked into the LongRunning option and I can't see how that would be hanging the process shutdown, but it would be good to get a memory dump from the customer if we can @mauroservienti. Which case was this?

**Comment by mauroservienti** [comment]
#17438 @colin-higgins

**Comment by mikeminutillo** [comment]
Hey all. I spent some time today trying to figure out the exact status of this issue just based on the associated comments and PRs. Can someone read and confirm the following?

- This bug was raised against v6 but there is an associated [fix for the current prod version](https://github.com/Particular/NServiceBus.RavenDB/issues/173). That fix has not been released yet?
- There was a [PR proposing a fix](https://github.com/Particular/NServiceBus.RavenDB/pull/165) which got moved to [this PR](https://github.com/Particular/NServiceBus.RavenDB/pull/166) which led to a [proposed change in the core](https://github.com/Particular/NServiceBus/pull/3497) which was cancelled in favor of [this spike](https://github.com/Particular/NServiceBus/issues/3532) which has not yet been started. 

Is this issue effectively blocked waiting for the spike? If not, what is required to get this issue closed? The listed Plan of Attack seems to be very out of date with what is going on here.

**Comment by tmasternak** [comment]
@mauroservienti  If the spike has been a blocker it should be removed now (see: https://github.com/Particular/NServiceBus/issues/3532). If that's indeed the case can we continue work on this issue?

**Comment by mauroservienti** [comment]
No blockers here, @colin-higgins is actively working on this and we are at the testing phase

**Comment by mauroservienti** [comment]
As discussed we decided to have this included in vNext

**Comment by colin-higgins** [comment]
To clarify why we decided to not include this in vNext or release the patch. The customer who reported the issue found the problem hanging the endpoint in their own code. So we are fixing a theoretical scenario that hasn't actually cropped up yet. If we want to get rid of the streaming API we can still take the paging solution in the spike and apply it to vNext, but it seems YAGNI to me.

**Comment by asednev** [comment]
@colin-higgins I am one of the customers experiencing this issue. We are running into this problem while using NServiceBus.Host to host our NServiceBus application. I will be glad to answer any of your questions in addition to what I provided to support.

**Comment by DavidBoike** [comment]
@asednev Fantastic! I remember you from NSBCon. We'll look up the details of your case and reassess. On my phone so hard to look right now if you've already done this, but would it be possible to get a memory dump when the endpoint hangs? That would help us to ensure we have the right solution to the right problem.

**Comment by asednev** [comment]
@DavidBoike, I remember you from NSBCon as well :) It was a great conference. There was an email thread with support with subject "NSB occasionally stops processing defers". Andreas analyzed memory dumps and arrived to conclusions about locking in RavenDB. In addition to issues with defers, we discussed issue with RavenDB preventing service from stopping. Please let me know if you still need dumps. I can try to locate old ones.

**Comment by andreasohlund** [comment]
I analyzed those dumps, I can confirm that threads are hanging on the streaming call and prevents shutdown.

**Comment by mauroservienti** [comment]
this allows the streaming query to be cancelled if the CancellationToken is signalled, is this going to be enough @andreasohlund and @asednev or what hangs is the internal behavior of the streaming engine so if it happens we are still stuck in a deadlock?

**Comment by DavidBoike** [comment]
The fix we can roll into 3.0.7 also removes the use of the streaming API in favor of standard queries. We're suspicious that an implementation detail of the streaming API is responsible for the hang and there's no reason we *have* to use it.

**Comment by DavidBoike** [comment]
This will be released with 3.0.7, updating milestone to match.
